src/io/flutter/editor/ActiveEditorsOutlineService.java

```diff
@@ -186,8 +189,61 @@ public FlutterOutline get(String path) {
    * Gets the most up-to-date {@link FlutterOutline} for the {@param file}.
    */
   @Nullable
-  public FlutterOutline get(VirtualFile file) {
-    return get(file.getCanonicalPath());
+  public FlutterOutline get(@NotNull VirtualFile file) {
+    return pathToOutline.get(file.getCanonicalPath());
+  }
+
+  /**
+   * Gets the up-to-date {@link FlutterOutline} for the {@param file}.
+   *
+   * <p>
+   * If the cached outline is not updated yet, will request an updated outline from the analysis service.
+   */
+  public FlutterOutline request(@NotNull VirtualFile file) {
+    if (isUpToDate(file)) {
+      return get(file);
+    }
+    final String path = file.getCanonicalPath();
+    // If we haven't gotten any data, set up a temporary listener for this file, and wait for a response.
+    // We block to wait with a Lock.
+    final Lock lock = new ReentrantLock();
+    final Condition notLoaded = lock.newCondition();
+    lock.lock();
+    final Listener listener = (p, o) -> {
+      if (p.equals(path)) {
+        notLoaded.signal();
+      }
+    };
+    addListener(listener);
+    try {
+      notLoaded.await(TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+    }
+    catch (InterruptedException ignored) {
+      // The thread was interrupted while waiting for the load.
+    }
+    finally {
+      lock.unlock();
+      removeListener(listener);
+    }
+    // Return the new value, which either:
+    //  - is updated, and the listener unlocked the request
+    //  - is outdated but the listener timed out waiting for a response.
+    return get(path);
+  }
+
+  private boolean isUpToDate(@NotNull VirtualFile file) {
+    final FlutterOutline outline = get(file);
+    final PsiFile psiFile = getPsiFile(file);
+    final DartAnalysisServerService das = DartAnalysisServerService.getInstance(project);
+    return psiFile != null && outline != null && (
+      psiFile.getTextLength() == outline.getLength()
+      || psiFile.getTextLength() == das.getConvertedOffset(file, outline.getLength())
+    );
+  }
+
+  @VisibleForTesting
+  protected PsiFile getPsiFile(VirtualFile file) {
+    return PsiManager.getInstance(project).findFile(file);
```

Comment:
Make this a utility method on a different class if you need it.

Reply:
It's no longer necessary, but it was originally intended for use during tests to inject custom files.